### PR TITLE
Add dynamically adjust the upper limit of the data that each layer configure option.

### PR DESCRIPTION
**Pull request by WyattJia**
* Levels' Target Size [level_compaction_dynamic_level_bytes](https://github.com/facebook/rocksdb/wiki/Leveled-Compaction#levels-target-size)
* Reference issue: https://github.com/KvrocksLabs/kvrocks/issues/488

Please check and review, thanks.

### CODE REVIEW

**Comment by git-hulk** [comment]
Cool, thanks for your contribution.

**Comment by git-hulk** [comment]
@WyattJia I have a look at the PR, overall was good to me. But we need to fix those lint issues first:
```
src/config.cc:152:  Lines should be <= 120 characters long  [whitespace/line_length] [2]
src/config.cc:153:  Lines should be <= 120 characters long  [whitespace/line_length] [2]
src/config.cc:427:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/config.cc:428:  Lines should be <= 120 characters long  [whitespace/line_length] [2]
```

**Comment by WyattJia** [comment]
> @WyattJia I have a look at the PR, overall was good to me. But we need to fix those lint issues first:
> 
> ```
> src/config.cc:152:  Lines should be <= 120 characters long  [whitespace/line_length] [2]
> src/config.cc:153:  Lines should be <= 120 characters long  [whitespace/line_length] [2]
> src/config.cc:427:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
> src/config.cc:428:  Lines should be <= 120 characters long  [whitespace/line_length] [2]
> ```

Copy that, I've formated and pushed code, thanks for your review.

File: kvrocks.conf [hunk 1]
```diff
@@ -597,6 +597,49 @@ rocksdb.enable_blob_garbage_collection yes
...
+# With this option on, from an empty DB, we make last level the base level,
+# which means merging L0 data into the last level, until it exceeds
+# max_bytes_for_level_base. And then we make the second last level to be
+# base level, to start to merge L0 data to second last level, with its
+# target size to be 1/max_bytes_for_level_multiplier of the last level's
+# extra size. After the data accumulates more so that we need to move the
+# base level to the third last one, and so on.
+#
+# Default False
+rocksdb.level_compaction_dynamic_level_bytes false
```
Comment (caipengbo) [inline kvrocks.conf]: Can we simplify this configuration item comment?
I think we should add a reference address(how about rocksdb wiki?) instead of copying rocksdb's comments, which are more detailed.

File: src/config.h [hunk 2]
```diff
@@ -130,6 +130,9 @@ struct Config{
     int blob_file_size;
     bool enable_blob_garbage_collection;
     int blob_garbage_collection_age_cutoff;
+    int max_bytes_for_level_base;
+    int max_bytes_for_level_multiplier;
```
Comment (caipengbo) [inline src/config.h:134]: This configuration item is a `double` in rocksdb, but `int` type in kvrocks, so we should specify it in kvrocks.conf(it would be perfect to support `double`).

File: src/config.cc [hunk 3]
```diff
@@ -404,6 +411,30 @@ void Config::initFieldCallback() {
...
         return srv->storage_->SetColumnFamilyOption(trimRocksDBPrefix(k), std::to_string(cutoff));
       }},
+      {"rocksdb.level_compaction_dynamic_level_bytes", [this](Server* srv, const std::string &k,
+                                                        const std::string& v)->Status {
+        if (!srv) return Status::OK();
+        if (!RocksDB.level_compaction_dynamic_level_bytes) {
+          return Status(Status::NotOK, errNotSetLevelCompactionDynamicLevelBytes);
+        }
+        std::string level_compaction_dynamic_level_bytes = v == "yes" ? "true" : "false";
+        return srv->storage_->SetColumnFamilyOption(trimRocksDBPrefix(k), level_compaction_dynamic_level_bytes);
```
Comment (caipengbo) [inline src/config.cc]: Does this option support dynamically changeable through SetOptions() API？I can not find any infomations here: https://github.com/facebook/rocksdb/blob/main/include/rocksdb/advanced_options.h#L594

File: src/config.cc [hunk 4]
```diff
@@ -145,7 +147,12 @@ Config::Config() {
       {"rocksdb.blob_file_size", false, new IntField(&RocksDB.blob_file_size, 128, 0, INT_MAX)},
       {"rocksdb.enable_blob_garbage_collection", false, new YesNoField(&RocksDB.enable_blob_garbage_collection, true)},
       {"rocksdb.blob_garbage_collection_age_cutoff",
-       false, new IntField(&RocksDB.blob_garbage_collection_age_cutoff, 25, 0, 100)}
+       false, new IntField(&RocksDB.blob_garbage_collection_age_cutoff, 25, 0, 100)},
+      {"rocksdb.max_bytes_for_level_base", false, new IntField(&RocksDB.max_bytes_for_level_base, 256, 0, INT16_MAX)},
+      {"rocksdb.max_bytes_for_level_multiplier",
```
Comment (git-hulk) [inline src/config.cc]:
```suggestion
      {"rocksdb.max_bytes_for_level_multiplier",
```

File: src/config.cc [hunk 5]
```diff
@@ -145,7 +147,12 @@ Config::Config() {
       {"rocksdb.blob_file_size", false, new IntField(&RocksDB.blob_file_size, 128, 0, INT_MAX)},
       {"rocksdb.enable_blob_garbage_collection", false, new YesNoField(&RocksDB.enable_blob_garbage_collection, true)},
       {"rocksdb.blob_garbage_collection_age_cutoff",
-       false, new IntField(&RocksDB.blob_garbage_collection_age_cutoff, 25, 0, 100)}
+       false, new IntField(&RocksDB.blob_garbage_collection_age_cutoff, 25, 0, 100)},
+      {"rocksdb.max_bytes_for_level_base", false, new IntField(&RocksDB.max_bytes_for_level_base, 256, 0, INT16_MAX)},
+      {"rocksdb.max_bytes_for_level_multiplier", 
+        false, new IntField(&RocksDB.max_bytes_for_level_multiplier, 10, 1, 100)},
+      {"rocksdb.level_compaction_dynamic_level_bytes",
```
Comment (git-hulk) [inline src/config.cc]:
```suggestion
      {"rocksdb.level_compaction_dynamic_level_bytes",
```

File: src/config.h (hunk 2 above)
Comment (WyattJia) [reply src/config.h:134]: Yep, I've add date type specify about `max_bytes_for_level_multiplier ` option in `kvrocks.conf`, thanks.

File: src/config.cc (hunk 3 above)
Comment (WyattJia) [reply src/config.cc]: > Does this option support dynamically changeable through SetOptions() API？I can not find any infomations here: https://github.com/facebook/rocksdb/blob/main/include/rocksdb/advanced_options.h#L594

I haven't found too, I changed this option load logic. Please help me review and check code, thanks.

File: kvrocks.conf (hunk 1 above)
Comment (WyattJia) [reply kvrocks.conf]: > Can we simplify this configuration item comment? I think we should add a reference address(how about rocksdb wiki?) instead of copying rocksdb's comments, which are more detailed.

I've rewrited comments, thanks!

File: kvrocks.conf [hunk 6]
```diff
@@ -597,6 +597,28 @@ rocksdb.enable_blob_garbage_collection yes
...
+
+# The purpose of this option is to dynamically adjust the upper limit of
+# the data that each layer can store according to the size of the different 
+# layers of the LSM. Enabling this option will bring some improvements in
+# deletion efficiency and space amplification, but it will lose a certain 
+# amount of read performance.
+# If you want know more detail about Levels' Target Size, you can read RocksDB wiki: 
+# https://github.com/facebook/rocksdb/wiki/Leveled-Compaction#levels-target-size
+#
+# Default False
```
Comment (caipengbo) [inline kvrocks.conf]:
```suggestion
# Default: false
```

File: kvrocks.conf [hunk 7]
```diff
@@ -597,6 +597,28 @@ rocksdb.enable_blob_garbage_collection yes
...
+rocksdb.level_compaction_dynamic_level_bytes false
+
+# Default 256 M
```
Comment (caipengbo) [inline kvrocks.conf]:
```suggestion
# Default: 256 M
```

File: kvrocks.conf [hunk 8]
```diff
@@ -597,6 +597,28 @@ rocksdb.enable_blob_garbage_collection yes
...
+rocksdb.max_bytes_for_level_base 256
+
+# This configuration value is a double type number in RocksDB, 
+# but kvrocks is not support double data type number yet,
+# so we use int data number instead of double currently.
+#
+# Default 10
```
Comment (caipengbo) [inline kvrocks.conf]:
```suggestion
# Default: 10
```

**Review by caipengbo — DISMISSED** [review]

**Review by caipengbo — DISMISSED** [review]
Good job!

File: src/config.cc [hunk 9]
```diff
@@ -21,6 +21,9 @@ const char *kDefaultNamespace = "__namespace";
 
 const char *errNotEnableBlobDB = "Must set rocksdb.enable_blob_files to yes first.";
 
+const char *errNotSetLevelCompactionDynamicLevelBytes =
```
Comment (git-hulk) [inline src/config.cc]:
```suggestion
const char *errNotSetLevelCompactionDynamicLevelBytes =
```

**Review by git-hulk — DISMISSED** [review]

File: src/config.cc [hunk 10]
```diff
@@ -145,7 +148,12 @@ Config::Config() {
       {"rocksdb.blob_file_size", false, new IntField(&RocksDB.blob_file_size, 128, 0, INT_MAX)},
       {"rocksdb.enable_blob_garbage_collection", false, new YesNoField(&RocksDB.enable_blob_garbage_collection, true)},
       {"rocksdb.blob_garbage_collection_age_cutoff",
-       false, new IntField(&RocksDB.blob_garbage_collection_age_cutoff, 25, 0, 100)}
+       false, new IntField(&RocksDB.blob_garbage_collection_age_cutoff, 25, 0, 100)},
+      {"rocksdb.max_bytes_for_level_base", false, new IntField(&RocksDB.max_bytes_for_level_base, 256, 0, INT16_MAX)},
```
Comment (git-hulk) [inline src/config.cc]: Default value should be 256MiB

File: src/config.cc [hunk 11]
```diff
@@ -145,7 +150,12 @@ Config::Config() {
...
```
Comment (git-hulk) [inline src/config.cc]:
```suggestion
      {"rocksdb.max_bytes_for_level_base", false, new IntField(&RocksDB.max_bytes_for_level_base, 268435456, 0, INT16_MAX)},
```

File: kvrocks.conf [hunk 12]
```diff
@@ -578,7 +578,7 @@ rocksdb.min_blob_size 4096
 # opened once this limit is reached.
 #
 # Default: 128 M
-rocksdb.blob_file_size 128
+rocksdb.blob_file_size 134217728
```
Comment (git-hulk) [inline kvrocks.conf]: I think we can submit another PR to resolve this issue? also need to change the default value in config.cc(`min_blob_size` was the same problem).

File: tests/config_test.cc [hunk 13]
```diff
@@ -54,6 +54,9 @@ TEST(Config, GetAndSet) {
       {"rocksdb.blob_file_size", "128"},
       {"rocksdb.enable_blob_garbage_collection", "yes"},
       {"rocksdb.blob_garbage_collection_age_cutoff", "25"},
+      {"rocksdb.max_bytes_for_level_base", "128"},
```
Comment (git-hulk) [inline tests/config_test.cc]:
```suggestion
      {"rocksdb.max_bytes_for_level_base", "268435456"},
```

File: src/config.cc [hunk 14]
```diff
@@ -145,7 +149,12 @@ Config::Config() {
       {"rocksdb.blob_file_size", false, new IntField(&RocksDB.blob_file_size, 128, 0, INT_MAX)},
       {"rocksdb.enable_blob_garbage_collection", false, new YesNoField(&RocksDB.enable_blob_garbage_collection, true)},
       {"rocksdb.blob_garbage_collection_age_cutoff",
-       false, new IntField(&RocksDB.blob_garbage_collection_age_cutoff, 25, 0, 100)}
+       false, new IntField(&RocksDB.blob_garbage_collection_age_cutoff, 25, 0, 100)},
+      {"rocksdb.max_bytes_for_level_base", false, new IntField(&RocksDB.max_bytes_for_level_base, 268435456, 0, INT16_MAX)},
```
Comment (git-hulk) [inline src/config.cc]:
```suggestion
      {"rocksdb.max_bytes_for_level_base",
        false, new IntField(&RocksDB.max_bytes_for_level_base, 268435456, 0, INT16_MAX)},
```

File: src/config.cc [hunk 15]
```diff
@@ -145,7 +149,13 @@ Config::Config() {
       {"rocksdb.blob_file_size", false, new IntField(&RocksDB.blob_file_size, 128, 0, INT_MAX)},
       {"rocksdb.enable_blob_garbage_collection", false, new YesNoField(&RocksDB.enable_blob_garbage_collection, true)},
       {"rocksdb.blob_garbage_collection_age_cutoff",
-       false, new IntField(&RocksDB.blob_garbage_collection_age_cutoff, 25, 0, 100)}
+       false, new IntField(&RocksDB.blob_garbage_collection_age_cutoff, 25, 0, 100)},
+      {"rocksdb.max_bytes_for_level_base",
+        false, new IntField(&RocksDB.max_bytes_for_level_base, 268435456, 0, INT16_MAX)},
```
Comment (git-hulk) [inline src/config.cc]:
```suggestion
        false, new IntField(&RocksDB.max_bytes_for_level_base, 268435456, 0, INT_MAX)},
```

**Review by git-hulk — DISMISSED** [review]

File: .gitignore [hunk 16]
```diff
@@ -43,4 +43,7 @@ version.h
 Makefile.dep
 make_config.mk
 .vscode
-kvrocks2redis
\ No newline at end of file
+kvrocks2redis
+
+src/compactdb/
+src/testsdb/
```
Comment (ShooterIT) [inline .gitignore]:
```suggestion
```
why did you add these, to avoid your tests polluting git repo? i think it is not necessary if just because of tests

Comment (WyattJia) [reply .gitignore]: Those are default generated test database pathes when I run kvrocks on project path, I guess other developers will generate those test databases on `kvrocks/src` directory too, so I ignore those pathes for develop convenience.

File: src/config.cc [hunk 17]
```diff
@@ -404,6 +414,30 @@ void Config::initFieldCallback() {
...
+        }
+        return srv->storage_->SetColumnFamilyOption(trimRocksDBPrefix(k), v);
+      }},
+      {"rocksdb.level_compaction_dynamic_level_bytes", [this](Server* srv, const std::string &k,
+                                                        const std::string& v)->Status {
+        if (!srv) return Status::OK();
+        if (!RocksDB.level_compaction_dynamic_level_bytes) {
+          return Status(Status::NotOK, errNotSetLevelCompactionDynamicLevelBytes);
+        }
+        std::string enable_blob_garbage_collection = v == "yes" ? "true" : "false";
```
Comment (ShooterIT) [inline src/config.cc]: you want to set `level_compaction_dynamic_level_bytes`, but use `enable_blob_garbage_collection` variable ?

File: kvrocks.conf [hunk 18]
```diff
@@ -597,6 +597,30 @@ rocksdb.enable_blob_garbage_collection yes
...
+# If you want know more detail about Levels' Target Size, you can read RocksDB wiki: 
+# https://github.com/facebook/rocksdb/wiki/Leveled-Compaction#levels-target-size
+#
+# Default: no
+rocksdb.level_compaction_dynamic_level_bytes no
+
+# The total file size of level-1 sst.
+#
+# Default: 256 M
+rocksdb.max_bytes_for_level_base 268435456
```
Comment (ShooterIT) [inline kvrocks.conf]: I think `MB` is much readable, even lose flexibility, and currently, we usually use `MB` too, @git-hulk WDYT?

Comment (git-hulk) [reply kvrocks.conf]: It should be fine to keep the same as RocksDB since we expected only a few users would adjust those options.

Comment (WyattJia) [reply kvrocks.conf]: @git-hulk @ShooterIT I'd updated max_bytes_for_level_base unit bytes to MiB.

**Comment by git-hulk** [comment]
cool, also need to multiple the `MiB` in config's callback like RocksDB.write_buffer_size

https://github.com/KvrocksLabs/kvrocks/blob/unstable/src/config.cc#L346

**Comment by WyattJia** [comment]
> cool, also need to multiple the `MiB` in config's callback like RocksDB.write_buffer_size
> 
> https://github.com/KvrocksLabs/kvrocks/blob/unstable/src/config.cc#L346

Done and thank you for your reminder.

**Comment by git-hulk** [comment]
@WyattJia Cool, really thanks for your contribution and patient.

**Comment by git-hulk** [comment]
@ShooterIT Please take a look again

**Review by ShooterIT — APPROVED** [review]
LGTM

**Review by git-hulk — APPROVED** [review]

File: src/config.cc [hunk 19]
```diff
@@ -404,6 +414,31 @@ void Config::initFieldCallback() {
...
+          return Status(Status::NotOK, errNotSetLevelCompactionDynamicLevelBytes);
+        }
+        return srv->storage_->SetColumnFamilyOption(trimRocksDBPrefix(k), v);
+      }},
+      {"rocksdb.level_compaction_dynamic_level_bytes", [this](Server* srv, const std::string &k,
+                                                        const std::string& v)->Status {
+        if (!srv) return Status::OK();
+        if (!RocksDB.level_compaction_dynamic_level_bytes) {
+          return Status(Status::NotOK, errNotSetLevelCompactionDynamicLevelBytes);
+        }
```
Comment (ShooterIT) [inline src/config.cc:438]: this judgement is necessary? a mistake?

File: kvrocks.conf [hunk 20]
```diff
@@ -597,6 +597,30 @@ rocksdb.enable_blob_garbage_collection yes
...
+# If you want know more details about Levels' Target Size, you can read RocksDB wiki: 
+# https://github.com/facebook/rocksdb/wiki/Leveled-Compaction#levels-target-size
+#
+# Default: no
+rocksdb.level_compaction_dynamic_level_bytes no
+
+# The total file size of level-1 sst.
+#
+# Default: 256 M
+rocksdb.max_bytes_for_level_base 256
```
Comment (ShooterIT) [inline kvrocks.conf:615]: since this config has `bytes` word, we should use byte unit, right? or we rename it to level_base_size? maybe rocksdb name is better?

This commit is not released in newest version, we have chances to change.

**Comment by ShooterIT** [comment]
@WyattJia have a look?

File: kvrocks.conf (hunk 20 above)
Comment (caipengbo) [reply kvrocks.conf:615]: Yeah, I also think `bytes` is better.
